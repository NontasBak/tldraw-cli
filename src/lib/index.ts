export { tldrawToImage } from './tldraw-to-image'
