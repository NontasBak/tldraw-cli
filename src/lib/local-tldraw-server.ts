import express from 'express'
import getPort from 'get-port'
import { type Server } from 'node:http'
import { type AddressInfo } from 'node:net'
import path, { dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

export default class LocalTldrawServer {
	verbose: boolean
	// eslint-disable-next-line perfectionist/sort-classes
	private server?: Server

	constructor(verbose = false) {
		this.verbose = verbose
	}

	close(): void {
		if (!this.server) throw new Error('Server not started')
		this.server.close()
		if (this.verbose) console.log('Stopped tldraw server')
	}

	async start() {
		// Serve local tldraw
		if (this.verbose) console.log('Starting tldraw server...')

		const scriptDirectory = dirname(fileURLToPath(import.meta.url))

		// Handle dev or prod relative paths, brittle
		const tldrawPath = path.join(
			scriptDirectory,
			scriptDirectory.endsWith('/src/lib')
				? '../../dist/tldraw'
				: scriptDirectory.endsWith('/dist/lib')
					? '../tldraw'
					: '../dist/tldraw',
		)

		if (this.verbose) console.log(`tldraw served from "${tldrawPath}"`)
		const app = express()
		const port = await getPort()
		app.use(express.static(tldrawPath))

		try {
			this.server = app.listen(port)
		} catch (error) {
			console.error(error)
		}

		if (this.verbose) console.log(`tldraw hosted at "${this.href}"`)
	}

	get href(): string {
		if (!this.server) throw new Error('Server not started')
		const { port } = this.server.address() as AddressInfo
		return `http://localhost:${port}`
	}
}
