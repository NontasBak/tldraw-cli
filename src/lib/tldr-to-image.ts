import type { ExportFormat } from '../types'
import express from 'express'
import getPort from 'get-port'
import fs from 'node:fs/promises'
import { type Server } from 'node:http'
import { type AddressInfo } from 'node:net'
import os from 'node:os'
import path from 'node:path'
import { dirname } from 'node:path'
import { fileURLToPath } from 'node:url'
import puppeteer from 'puppeteer'
import type { CDPSession } from 'puppeteer'
import untildify from 'untildify'

async function startServer(path: string): Promise<Server> {
	const app = express()
	const port = await getPort()
	app.use(express.static(path))

	return new Promise((resolve, reject) => {
		const server = app.listen(port, () => {
			resolve(server)
		})

		server.on('error', (error) => {
			reject(error)
		})
	})
}

async function waitForDownloadCompletion(client: CDPSession): Promise<string> {
	return new Promise((resolve, reject) => {
		client.on('Browser.downloadProgress', (event) => {
			if (event.state === 'completed') {
				resolve(event.guid)
			} else if (event.state === 'canceled') {
				reject(new Error('Download was canceled'))
			}
		})
	})
}

export async function tldrToImage(
	tldrPath: string,
	format: ExportFormat = 'svg',
	destination = './',
	verbose = false,
): Promise<string> {
	const scriptDirectory = dirname(fileURLToPath(import.meta.url))
	const resolvedTldrPath = path.resolve(untildify(tldrPath))
	if (verbose) console.log(`Loading tldr file "${resolvedTldrPath}"`)
	const tldrFile = await fs.readFile(resolvedTldrPath, 'utf8')

	// Serve local tldraw
	if (verbose) console.log('Starting tldraw server...')

	// Handle dev or prod relative paths, brittle
	console.log(`scriptDirectory: ${scriptDirectory}`)
	const tldrawPath = path.join(
		scriptDirectory,
		scriptDirectory.endsWith('/src/lib')
			? '../../dist/tldraw'
			: scriptDirectory.endsWith('/dist/lib')
				? '../tldraw'
				: '../dist/tldraw',
	)

	if (verbose) console.log(`tldraw served from "${tldrawPath}"`)

	const server = await startServer(tldrawPath)
	const { port } = server.address() as AddressInfo

	if (verbose) console.log(`tldraw hosted at "http://localhost:${port}"`)

	// Launch Puppeteer and access the Vite-served website
	if (verbose) console.log('Starting Puppeteer...')
	const browser = await puppeteer.launch({ headless: 'new' })
	const page = await browser.newPage()

	// Hush the favicon kvetch
	await page.setRequestInterception(true)
	page.on('request', (request) => {
		if (request.url().endsWith('favicon.ico')) {
			void request.respond({ status: 200 })
		} else {
			void request.continue()
		}
	})

	// Forward messages from the browser console
	page.on('console', (message) => {
		const messageType = message.type()
		const messageText = message.text()

		if (messageType === 'error') {
			console.error(`[Browser] ${messageText}`)
		} else if (messageType === 'warning') {
			console.warn(`[Browser] ${messageText}`)
		} else if (verbose) {
			console.log(`[Browser] ${messageText}`)
		}
	})

	const client = await page.target().createCDPSession()
	await client.send('Browser.setDownloadBehavior', {
		behavior: 'allowAndName',
		downloadPath: os.tmpdir(),
		eventsEnabled: true,
	})

	await page.goto(`http://localhost:${port}`)
	await page.waitForFunction(() => window.tldrawExportFile !== undefined)

	// Send the tldr file to the page
	if (verbose) console.log('Requesting download')
	await page.evaluate(
		(tldrFile, format) => {
			window.tldrawExportFile(tldrFile, format)
		},
		tldrFile,
		format,
	)

	const downloadGuid = await waitForDownloadCompletion(client)
	if (verbose) console.log('Download complete')

	await browser.close()
	if (verbose) console.log('Stopped Puppeteer')

	// Move and rename the downloaded file from temp to output destination
	const originalFilename = path.basename(tldrPath, path.extname(tldrPath))
	const downloadPath = path.join(os.tmpdir(), downloadGuid)
	const destinationPath = untildify(path.join(destination, `${originalFilename}.${format}`))
	await fs.rename(downloadPath, destinationPath)

	if (verbose) console.log(`Saved to "${destinationPath}"`)

	server.close()
	if (verbose) console.log('Stopped tldraw server')

	return path.resolve(destinationPath)
}
