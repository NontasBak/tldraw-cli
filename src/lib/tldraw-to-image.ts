/* eslint-disable complexity */

import type { ExportFormat } from '../types'
import express from 'express'
import getPort from 'get-port'
import fs from 'node:fs/promises'
import { type Server } from 'node:http'
import { type AddressInfo } from 'node:net'
import os from 'node:os'
import path from 'node:path'
import { dirname } from 'node:path'
import { fileURLToPath } from 'node:url'
import puppeteer from 'puppeteer'
import type { CDPSession, Page } from 'puppeteer'
import untildify from 'untildify'

export type TldrawImageOptions = {
	darkMode?: boolean
	format?: ExportFormat
	output?: string
	transparent?: boolean
	verbose?: boolean
}

const defaultOptions: Required<TldrawImageOptions> = {
	darkMode: false,
	format: 'svg',
	output: './',
	transparent: false,
	verbose: true,
}

export async function tldrawToImage(
	tldrPathOrUrl: string,
	options: TldrawImageOptions = {},
): Promise<string> {
	const resolvedOptions = { ...defaultOptions, ...stripUndefined(options) }
	const { darkMode, format, output, transparent, verbose } = resolvedOptions

	// Identify URL vs. file path
	const isLocal = !tldrPathOrUrl.startsWith('https://www.tldraw.com/')
	if (verbose) console.log(isLocal ? 'Local file detected' : 'tldraw URL detected')

	// Set up local server if appropriate
	let port = 0
	let server: Server | undefined
	if (isLocal) {
		// Serve local tldraw
		if (verbose) console.log('Starting tldraw server...')

		const scriptDirectory = dirname(fileURLToPath(import.meta.url))

		// Handle dev or prod relative paths, brittle
		const tldrawPath = path.join(
			scriptDirectory,
			scriptDirectory.endsWith('/src/lib')
				? '../../dist/tldraw'
				: scriptDirectory.endsWith('/dist/lib')
					? '../tldraw'
					: '../dist/tldraw',
		)

		if (verbose) console.log(`tldraw served from "${tldrawPath}"`)
		server = await startServer(tldrawPath)
		port = (server.address() as AddressInfo).port
		if (verbose) console.log(`tldraw hosted at "http://localhost:${port}"`)
	}

	// Set up Puppeteer
	if (verbose) console.log('Starting Puppeteer...')
	const browser = await puppeteer.launch({
		args: isLocal ? ['--no-sandbox', '--disable-web-security'] : [],
		headless: 'new',
	})
	const page = await browser.newPage()
	echoBrowserConsole(page, verbose)
	const client = await page.target().createCDPSession()
	await client.send('Browser.setDownloadBehavior', {
		behavior: 'allowAndName',
		downloadPath: os.tmpdir(),
		eventsEnabled: true,
	})

	if (isLocal) {
		// Hush the favicon kvetch (for local tldraw)
		await page.setRequestInterception(true)
		page.on('request', (request) => {
			if (request.url().endsWith('favicon.ico')) {
				void request.respond({ status: 200 })
			} else {
				void request.continue()
			}
		})

		// Load tldr data and put it in local storage
		const resolvedTldrPath = path.resolve(untildify(tldrPathOrUrl))
		if (verbose) console.log(`Loading tldr file "${resolvedTldrPath}"`)
		const tldrFile = await fs.readFile(resolvedTldrPath, 'utf8')

		await page.evaluateOnNewDocument((data) => {
			localStorage.clear()
			localStorage.setItem('tldrData', data)
		}, tldrFile)
	}

	// Navigate to tldraw
	const tldrawUrl = isLocal ? `http://localhost:${port}` : tldrPathOrUrl
	if (verbose) console.log(`Navigating to: ${tldrawUrl}`)
	await page.goto(tldrawUrl, { waitUntil: 'networkidle0' })

	// Set transparency
	if (verbose) console.log(`Setting background transparency: ${transparent}`)
	await setTransparency(page, transparent)

	// Set dark mode
	if (verbose) console.log(`Setting dark mode: ${darkMode}`)
	const originalDarkMode = await getDarkMode(page)
	await setDarkMode(page, darkMode)

	// Download
	// Brittle, TODO how to invoke this from the browser console?
	if (verbose) console.log('Requesting download')
	await closeMenus(page)
	await clickMenuTestIds(page, [
		'main.menu',
		'menu-item.edit',
		'menu-item.export-as',
		`menu-item.export-as-${format}`,
	])

	const downloadGuid = await waitForDownloadCompletion(client)
	if (verbose) console.log('Download complete')

	// Restore dark mode
	if (verbose) console.log(`Restoring dark mode: ${originalDarkMode}`)
	await setDarkMode(page, originalDarkMode)

	await browser.close()
	if (verbose) console.log('Stopped Puppeteer')

	if (isLocal && server) {
		server.close()
		if (verbose) console.log('Stopped tldraw server')
	}

	// TODO better naming strategy for URLs...
	// Move and rename the downloaded file from temp to output destination
	const outputFilename = isLocal
		? path.basename(tldrPathOrUrl, path.extname(tldrPathOrUrl))
		: downloadGuid
	const downloadPath = path.join(os.tmpdir(), downloadGuid)
	const outputPath = untildify(path.join(output, `${outputFilename}.${format}`))
	await fs.rename(downloadPath, outputPath)

	if (verbose) console.log(`Saved to "${outputPath}"`)
	return path.resolve(outputPath)
}

// Helpers

function echoBrowserConsole(page: Page, verbose: boolean) {
	page.on('console', (message) => {
		const messageType = message.type()
		const messageText = message.text()

		if (messageType === 'error') {
			console.error(`[Browser] ${messageText}`)
		} else if (messageType === 'warning') {
			console.warn(`[Browser] ${messageText}`)
		} else if (verbose) {
			console.log(`[Browser] ${messageText}`)
		}
	})
}

async function startServer(path: string): Promise<Server> {
	const app = express()
	const port = await getPort()
	app.use(express.static(path))

	return new Promise((resolve, reject) => {
		const server = app.listen(port, () => {
			resolve(server)
		})

		server.on('error', (error) => {
			reject(error)
		})
	})
}

async function waitForDownloadCompletion(client: CDPSession): Promise<string> {
	return new Promise((resolve, reject) => {
		client.on('Browser.downloadProgress', (event) => {
			if (event.state === 'completed') {
				resolve(event.guid)
			} else if (event.state === 'canceled') {
				reject(new Error('Download was canceled'))
			}
		})
	})
}

async function clickMenuTestIds(page: Page, testIds: string[]) {
	for (const testId of testIds) {
		await page.waitForSelector(`[data-testid="${testId}"]`)
		await page.click(`[data-testid="${testId}"]`)
	}
}

function stripUndefined(options: Record<string, unknown>): Record<string, unknown> {
	return Object.fromEntries(Object.entries(options).filter(([, value]) => value !== undefined))
}

async function closeMenus(page: Page): Promise<void> {
	await page.evaluate(`app.clearOpenMenus()`)
}

// TODO possible performance gains by only setting transparency as needed?
// async function getTransparency(page: Page): Promise<boolean> {
// 	return !(await page.evaluate('editor.getInstanceState().exportBackground'))
// }

async function getDarkMode(page: Page): Promise<boolean> {
	return Boolean(await page.evaluate('editor.user.getUserPreferences().isDarkMode'))
}

async function setDarkMode(page: Page, darkMode: boolean): Promise<void> {
	await page.evaluate(`editor.user.updateUserPreferences({ isDarkMode: ${darkMode}})`)
}

// Ephemeral means we don't have to restore the user's value
async function setTransparency(page: Page, transparent: boolean): Promise<void> {
	await page.evaluate(
		`editor.updateInstanceState(
			{ exportBackground: ${!transparent} },
			{ ephemeral: true },
		 )`,
	)
}
