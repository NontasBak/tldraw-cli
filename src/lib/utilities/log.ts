/* eslint-disable perfectionist/sort-objects */
// Staying basic, always log to stderr

import chalk from 'chalk'

const log = {
	verbose: false,

	info(...data: unknown[]): void {
		if (!this.verbose) return
		// Even info logs to stderr for ease of redirection
		console.warn(chalk.green('[Info]'), ...data)
	},
	infoPrefixed(prefix: string, ...data: unknown[]): void {
		this.info(chalk.blue(`[${prefix}]`), ...data)
	},

	warn(...data: unknown[]): void {
		console.warn(chalk.yellow('[Warning]'), ...data)
	},
	warnPrefixed(prefix: string, ...data: unknown[]): void {
		this.warn(chalk.blue(`[${prefix}]`), ...data)
	},

	error(...data: unknown[]): void {
		console.error(chalk.red('[Error]'), ...data)
	},
	errorPrefixed(prefix: string, ...data: unknown[]): void {
		this.error(chalk.blue(`[${prefix}]`), ...data)
	},
}

export default log
