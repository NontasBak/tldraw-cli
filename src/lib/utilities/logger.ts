// Staying basic

import chalk from 'chalk'

export let cli = false
export let verbose = false

export function setCli(value: boolean) {
	cli = value
}

export function setVerbose(value: boolean) {
	verbose = value
}

export function log(...data: unknown[]): void {
	// To stdout in all contexts
	console.log(...data)
}

export function warn(...data: unknown[]): void {
	// To stderr in all contexts
	console.warn(chalk.yellow('[Warning]'), ...data)
}

export function error(...data: unknown[]): void {
	// To stderr in all contexts
	console.error(chalk.red('[Error]'), ...data)
}

export function info(...data: unknown[]): void {
	if (!verbose) return
	// To stderr in cli only
	if (cli) {
		console.warn(chalk.green('[Info]'), ...data)
	} else {
		console.log(...data)
	}
}
