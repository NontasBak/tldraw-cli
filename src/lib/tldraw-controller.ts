// Note special inline IIFE import, see ./plugins/esbuild-plugin-iife.ts
import downloadTldrInlineScript from './inline/download-tldr?iife'
import uploadTldrInlineScript from './inline/upload-tldr?iife'
import type { TldrawFormat } from './tldraw-to-image'
import log from './utilities/log'
import slugify from '@sindresorhus/slugify'
import * as cheerio from 'cheerio'
import fs from 'node:fs/promises'
import os from 'node:os'
import path from 'node:path'
import puppeteer from 'puppeteer'
import type { Browser, CDPSession, Page } from 'puppeteer'
import untildify from 'untildify'

type PageFrame = { id: string; name: string }

/* eslint-disable perfectionist/sort-classes */
export default class TldrawController {
	private page?: Page
	private isEmpty?: boolean
	private browser?: Browser
	private client?: CDPSession
	private originalDarkMode?: boolean

	constructor(private readonly href: string) {
		this.href = href
	}

	private get isLocal(): boolean {
		return this.href.startsWith('http://localhost')
	}

	async start() {
		// Set up Puppeteer
		log.info('Starting Puppeteer...')
		this.browser = await puppeteer.launch({
			args: this.isLocal ? ['--no-sandbox', '--disable-web-security'] : [],
			headless: true,
		})

		this.page = await this.browser.newPage()

		// Set up console logging passthrough
		this.page.on('console', (message) => {
			const messageType = message.type()
			const messageText = message.text()

			if (messageType === 'error') {
				log.errorPrefixed('Browser', messageText)
			} else if (messageType === 'warn') {
				log.warnPrefixed('Browser', messageText)
			} else {
				log.infoPrefixed('Browser', messageText)
			}
		})

		// Set up download rules
		this.client = await this.page.target().createCDPSession()
		await this.client.send('Browser.setDownloadBehavior', {
			behavior: 'allowAndName',
			downloadPath: os.tmpdir(),
			eventsEnabled: true,
		})

		// Navigate to tldraw
		log.info(`Navigating to: ${this.href}`)
		await this.page.goto(this.href, { waitUntil: 'networkidle0' })

		// Check for emptiness
		const shapeCount = (await this.page.evaluate('editor.getCurrentPageShapes().length')) as number
		this.isEmpty = shapeCount === 0
	}

	async close() {
		if (!this.browser) throw new Error('Controller not started')
		if (this.originalDarkMode !== undefined) {
			log.info(`Restoring dark mode: ${this.originalDarkMode}`)
			await this.setDarkMode(this.originalDarkMode)
		}

		await this.browser.close()
		log.info('Stopped controller')
	}

	// Public method doesn't expose pageFrame
	async download(
		output: string,
		filename: string,
		format: TldrawFormat,
		stripStyle: boolean,
		print: boolean,
	): Promise<string[]> {
		return this._download(output, filename, format, stripStyle, undefined, print)
	}

	async downloadFrame(
		output: string,
		filename: string,
		format: TldrawFormat,
		stripStyle: boolean,
		frameNameOrId: string,
		print: boolean,
	): Promise<string[]> {
		return this.downloadFrames(output, filename, format, stripStyle, [frameNameOrId], print)
	}

	async downloadFrames(
		output: string,
		filename: string,
		format: TldrawFormat,
		stripStyle: boolean,
		frameNamesOrIds: string[],
		print: boolean,
	): Promise<string[]> {
		// Validate frame existence
		const validPageFrames: PageFrame[] = []

		for (const frame of frameNamesOrIds) {
			const pageFrame = await this.getPageFrameWithNameOrId(frame)
			if (pageFrame === undefined) {
				log.warn(`Frame "${frame}" not found, skipping`)
			} else {
				validPageFrames.push(pageFrame)
			}
		}

		if (validPageFrames.length === 0) {
			throw new Error('No valid frames found')
		}

		// Check for frame.name collisions, if we have any then include frame id in the filename
		const validFrameNames = validPageFrames.map((frame) => slugify(frame.name))
		const isFrameNameCollision = validFrameNames.length !== new Set(validFrameNames).size

		if (isFrameNameCollision) {
			log.warn(
				'Frame names are not unique, including frame IDs in the output filenames to avoid collisions',
			)
		}

		const outputAccumulator: string[] = []
		for (const frame of validPageFrames) {
			const frameSuffix = isFrameNameCollision ? `-${frame.id.replace('shape:', '')}` : ''
			outputAccumulator.push(
				...(await this._download(
					output,
					`${filename}${frameSuffix}`,
					format,
					stripStyle,
					frame,
					print,
				)),
			)
		}

		return outputAccumulator
	}

	async downloadAllFrames(
		output: string,
		filename: string,
		format: TldrawFormat,
		stripStyle: boolean,
		print: boolean,
	): Promise<string[]> {
		const pageFrames = await this.getPageFrames()
		const frameNamesOrIds = pageFrames.map((f) => f.id)
		return this.downloadFrames(output, filename, format, stripStyle, frameNamesOrIds, print)
	}

	// Ephemeral means we don't have to restore the user's value
	async setTransparency(transparent: boolean): Promise<void> {
		if (!this.page) throw new Error('Controller not started')
		log.info(`Setting background transparency: ${transparent}`)
		await this.page.evaluate(
			`editor.updateInstanceState(
			{ exportBackground: ${!transparent} },
			{ ephemeral: true },
		 )`,
		)
	}

	async setDarkMode(darkMode: boolean) {
		if (!this.page) throw new Error('Controller not started')
		log.info(`Setting dark mode: ${darkMode}`)
		if (!this.originalDarkMode) this.originalDarkMode = await this.getDarkMode()
		await this.page.evaluate(`editor.user.updateUserPreferences({ isDarkMode: ${darkMode}})`)
	}

	async loadFile(filePath: string) {
		if (!this.page) throw new Error('Controller not started')
		if (this.isLocal)
			throw new Error(
				'File loading is only supported for remote tldraw.com instances. See tldraw-open for local file loading approach.',
			)

		await this.closeMenus()
		await this.page.evaluate(`userPreferences.showFileOpenWarning.set(false);`)

		log.info(`Uploading local file to tldraw.com: ${filePath}`)
		const tldrFile = await fs.readFile(filePath, 'utf8')

		// We have to call a custom function to upload the tldr file,
		// puppeteer waitForFileChooser fileInput.accept etc. does NOT work
		await this.page.evaluate(uploadTldrInlineScript)
		await this.page.evaluate("console.log('window.uploadTldr')")
		await this.page.evaluate(`window.uploadTldr(${tldrFile.toString()})`)
	}

	async getShareUrl(): Promise<string> {
		if (!this.page) throw new Error('Controller not started')
		if (this.isLocal)
			throw new Error('Share URLs are only supported for remote tldraw.com instances.')

		await this.closeMenus()
		await this.clickMenuTestIds(['main.menu', 'menu-item.file', 'menu-item.share-project'])
		await this.page.waitForNavigation({ waitUntil: 'networkidle0' })

		// Clear search params with viewport offset
		// Tried zooming to fit and then waiting for param update,
		// but wasn't reliable
		const shareUrl = new URL(this.page.url())
		shareUrl.search = ''
		return shareUrl.href
	}

	// eslint-disable-next-line @typescript-eslint/naming-convention
	private async _download(
		output: string,
		filename: string,
		format: TldrawFormat,
		stripStyle: boolean,
		pageFrame: PageFrame | undefined, // Trust that existence has been validated
		print: boolean,
	): Promise<string[]> {
		if (!this.page) throw new Error('Controller not started')

		if (this.isEmpty) {
			throw new Error('Cannot export an empty document')
		}

		if (stripStyle && format !== 'svg') {
			log.warn('--strip-style is only supported for SVG output')
		}

		if (pageFrame !== undefined && format === 'tldr') {
			log.warn('--frames is not supported for tldr output, downloading entire document')
		}

		// Brittle, TODO how to invoke this from the browser console?
		const completionPromise = this.waitForDownloadCompletion()
		await this.closeMenus()

		let frameSuffix = ''

		// Todo what happens with tldr?
		if (pageFrame !== undefined && format !== 'tldr') {
			log.info(`Selecting sketch frame "${pageFrame.name}" with ID "${pageFrame.id}"`)

			frameSuffix = `-${slugify(pageFrame.name)}`

			// Select the frame shape
			await this.page.evaluate('editor.selectNone()')
			await this.page.evaluate(`editor.select('${pageFrame.id}')`)
		}

		if (pageFrame === undefined && format === 'json') {
			// For some reason json export returns undefined when nothing
			// is selected
			await this.page.evaluate('editor.selectAll()')
		}

		// eslint-disable-next-line unicorn/prefer-ternary
		if (format === 'tldr') {
			// We have to call a custom function to download the tldr file
			await this.page.evaluate(downloadTldrInlineScript)
		} else {
			// Export
			await this.clickMenuTestIds([
				'main.menu',
				'menu-item.edit',
				'menu-item.export-as',
				`menu-item.export-as-${format}`,
			])
		}

		const downloadGuid = await completionPromise

		// _really_ wait for download to complete, can get intermittent failures
		// without this
		await this.page.waitForNetworkIdle()

		// Move and rename the downloaded file from temp to output destination
		const downloadPath = path.join(os.tmpdir(), downloadGuid)

		// Don't move the file if we're printing
		const outputPath = print
			? downloadPath
			: path.resolve(untildify(path.join(output, `${filename}${frameSuffix}.${format}`)))

		if (!print) await fs.rename(downloadPath, outputPath)

		if (stripStyle && format === 'svg') {
			// Strip style from the SVG
			const svg = await fs.readFile(outputPath, 'utf8')
			const strippedSvg = this.stripStyleElement(svg)
			await fs.writeFile(outputPath, strippedSvg)
		}

		// Naive implementation
		if (print) {
			if (format === 'png') {
				// Convert to base64
				const buffer = await fs.readFile(outputPath)
				const outputBase64 = buffer.toString('base64')
				return [outputBase64]
			}

			// All others are plain text, without newlines
			const outputString: string = await fs.readFile(outputPath, 'utf8')
			const outputStringNoNewlines = outputString.replaceAll('\n', '')
			return [outputStringNoNewlines]
		}

		return [outputPath]
	}

	private async closeMenus(): Promise<void> {
		if (!this.page) throw new Error('Controller not started')
		await this.page.evaluate(`app.clearOpenMenus()`)
	}

	private async clickMenuTestIds(testIds: string[]) {
		if (!this.page) throw new Error('Controller not started')
		for (const testId of testIds) {
			await this.page.waitForSelector(`[data-testid="${testId}"]`)
			await this.page.click(`[data-testid="${testId}"]`)
		}
	}

	private async waitForDownloadCompletion(): Promise<string> {
		return new Promise((resolve, reject) => {
			if (!this.client) throw new Error('Controller not started')
			this.client.on('Browser.downloadProgress', (event) => {
				if (event.state === 'completed') {
					resolve(event.guid)
				} else if (event.state === 'canceled') {
					reject(new Error('Download was canceled'))
				}
			})
		})
	}

	// TODO memoize...
	private async getPageFrameWithNameOrId(nameOrId: string): Promise<PageFrame | undefined> {
		const pageFrames = await this.getPageFrames()

		if (pageFrames.length === 0) {
			throw new Error('No frames found')
		}

		return (
			pageFrames.find((f) => f.name === nameOrId || slugify(f.name) === nameOrId) ??
			pageFrames.find(
				(f) => f.id === (nameOrId.startsWith('shape:') ? nameOrId : `shape:${nameOrId}`),
			)
		)
	}

	// TODO possible performance gains by only setting transparency as needed?
	// async function getTransparency(page: Page): Promise<boolean> {
	// 	return !(await page.evaluate('editor.getInstanceState().exportBackground'))
	// }

	private async getDarkMode(): Promise<boolean> {
		if (!this.page) throw new Error('Controller not started')
		return Boolean(await this.page.evaluate('editor.user.getUserPreferences().isDarkMode'))
	}

	// TODO memoize
	private async getPageFrames(): Promise<PageFrame[]> {
		if (!this.page) throw new Error('Controller not started')
		// Returns empty array if no frames
		return (await this.page.evaluate(
			`editor.getCurrentPageShapes().reduce((accumulator, shape) => {
				if (shape.type === 'frame') {
					accumulator.push({ id: shape.id, name: shape.props.name })
				}
				return accumulator
			}, [])`,
		)) as PageFrame[]
	}

	private stripStyleElement(svg: string): string {
		const dom = cheerio.load(svg, { xmlMode: true })
		dom('style').remove()
		return dom.xml()
	}
}
