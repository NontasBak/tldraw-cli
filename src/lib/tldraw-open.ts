import LocalTldrawServer from './local-tldraw-server'
import { tldrawToImage } from './tldraw-to-image'
import log from './utilities/log'
import { validatePathOrUrl } from './validation'
import { nanoid } from 'nanoid'
import fs from 'node:fs/promises'
import os from 'node:os'
import open from 'open'

export async function tldrawOpen(tldrPathOrUrl?: string): Promise<string> {
	let tldrData: string | undefined

	if (tldrPathOrUrl !== undefined) {
		const validatedPathOrUrl = validatePathOrUrl(tldrPathOrUrl, {
			requireFileExistence: true,
			validFileExtensions: ['.tldr'],
			validHostnames: ['www.tldraw.com'],
		})

		// Download remote urls
		if (typeof validatedPathOrUrl !== 'string') {
			// Download to temp
			const [savedFile] = await tldrawToImage(validatedPathOrUrl.href, {
				format: 'tldr',
				name: nanoid(),
				output: os.tmpdir(),
			})
			const href = await tldrawOpen(savedFile)
			// Clean up the local file
			await fs.rm(savedFile, { force: true })
			return href
		}

		// Read file into base64 string

		// Start the server
		tldrData = await fs.readFile(validatedPathOrUrl, 'utf8')
	}

	const tldrawServer = new LocalTldrawServer(tldrData)
	await tldrawServer.start()

	// If running form the cli, we have to return the href before opening the
	// browser since it's long-running
	if (tldrPathOrUrl === undefined) {
		log.info(`Opened empty tldraw sketch at ${tldrawServer.href}`)
	} else {
		log.info(`Opened tldraw sketch "${tldrPathOrUrl}" at ${tldrawServer.href}`)
	}

	process.stdout.write(`${tldrawServer.href}\n`)

	// Open in default browser
	// Some tldraw functions require continued access to the server,
	// so it keeps running until all browsers are closed
	await open(tldrawServer.href, { wait: true })

	return tldrawServer.href
}
