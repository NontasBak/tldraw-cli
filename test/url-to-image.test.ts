import { tldrawToImage } from '../src/lib/tldraw-to-image'
import { expectFileToBeValid } from './utilities/file'
import { randomId } from './utilities/random'
import { mkdirSync, rmSync, rmdirSync } from 'node:fs'
import { expect, it } from 'vitest'

const cleanUp = true
const tldrawTestUrl = 'https://www.tldraw.com/s/v2_c_9nMYBwT8UQ99RGDWfGr8H'

it('should convert the tldraw url to an svg in the current folder by default', async () => {
	const savedImageFileName = await tldrawToImage(tldrawTestUrl)
	// TODO vet file names

	expectFileToBeValid(savedImageFileName, 'svg')

	if (cleanUp) rmSync(savedImageFileName)
})

it('should convert the tldraw url to an svg when specified', async () => {
	const savedImageFileName = await tldrawToImage(tldrawTestUrl, { format: 'svg' })
	// TODO vet file names

	expectFileToBeValid(savedImageFileName, 'svg')

	if (cleanUp) rmSync(savedImageFileName)
})

it('should convert the tldraw url to a png when specified', async () => {
	const savedImageFileName = await tldrawToImage(tldrawTestUrl, { format: 'png' })

	expectFileToBeValid(savedImageFileName, 'png')

	if (cleanUp) rmSync(savedImageFileName)
})

it('should save the file to a specific directory when specified', async () => {
	const randomPath = randomId()
	mkdirSync(randomPath)

	const savedImageFileName = await tldrawToImage(tldrawTestUrl, {
		format: 'png',
		output: `./${randomPath}/`,
	})
	expect(savedImageFileName).toContain(randomPath)

	expectFileToBeValid(savedImageFileName, 'png')

	if (cleanUp) rmdirSync(randomPath, { recursive: true })
})
