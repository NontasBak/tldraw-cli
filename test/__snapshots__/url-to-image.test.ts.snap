// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should convert the tldraw url to a png when specified 1`] = `"9fb348e37e574a11d5948b7c89d26479090e038383b7260e16018bec7b696b99"`;

exports[`should convert the tldraw url to an svg in the current folder by default 1`] = `"e48f33d96c7cf113b4fafed71bc8e38bce540eef9bf56d0e2a93adad9c3e9c2b"`;

exports[`should convert the tldraw url to an svg when specified 1`] = `"e48f33d96c7cf113b4fafed71bc8e38bce540eef9bf56d0e2a93adad9c3e9c2b"`;

exports[`should save the file to a specific directory when specified 1`] = `"9fb348e37e574a11d5948b7c89d26479090e038383b7260e16018bec7b696b99"`;
